src/main.c

```diff
@@ -110,41 +121,63 @@ struct {
   pthread_cond_t event_cv;
   pthread_mutex_t event_lock;
 
-  char workdir[_POSIX_PATH_MAX + 1];
-
-} zl_context = {0};
+  char instance_dir[PATH_MAX + 1];
+  char root_dir[PATH_MAX+1];
+  
+  enum zl_start_mode_t start_mode;
+  
+} zl_context = {.config = {.debug_mode = true}};
 
 #define INFO(fmt, ...)  printf("%s INFO:  "fmt, gettime().value, ##__VA_ARGS__)
 #define WARN(fmt, ...)  printf("%s WARN:  "fmt, gettime().value, ##__VA_ARGS__)
 #define DEBUG(fmt, ...) if (zl_context.config.debug_mode) \
   printf("%s DEBUG: "fmt, gettime().value, ##__VA_ARGS__)
 #define ERROR(fmt, ...) printf("%s ERROR: "fmt, gettime().value, ##__VA_ARGS__)
 
-static int init_context(const struct zl_config_t *cfg) {
+static int init_context(int argc, char **argv, const struct zl_config_t *cfg) {
 
-  const char *workdir = getenv("WORKDIR");
-  if (workdir == NULL) {
-    ERROR("WORKDIR env variable not found\n");
+  const char *instance_dir = getenv("INSTANCE_DIR");
+  if (instance_dir == NULL) {
+    ERROR("INSTANCE_DIR env variable not found\n");
     return -1;
   }
 
-  const char *dir_start = workdir;
-  const char *dir_end = dir_start + strlen(workdir) - 1;
+  const char *dir_start = instance_dir;
+  const char *dir_end = dir_start + strlen(instance_dir) - 1;
   while (*dir_end == ' ' && dir_end != dir_start) {
     dir_end--;
   }
 
   size_t dir_len = dir_end - dir_start + 1;
-  if (dir_len > sizeof(zl_context.workdir) - 1) {
-    ERROR("WORKDIR env too large\n");
+  if (dir_len > sizeof(zl_context.instance_dir) - 1) {
+    ERROR("INSTANCE_DIR env too large\n");
     return -1;
   }
-
-  memset(zl_context.workdir, 0, sizeof(zl_context.workdir));
-  memcpy(zl_context.workdir, dir_start, dir_len);
+  
+  memset(zl_context.instance_dir, 0, sizeof(zl_context.instance_dir));
+  memcpy(zl_context.instance_dir, dir_start, dir_len);
+  
+  char *root_dir = getenv("ROOT_DIR");
+  if (root_dir == NULL) {
+    ERROR("ROOT_DIR env variable not found\n");
+    return -1;
+  }
+  for (int i = strlen(root_dir) - 1; i >=0; i++) {
+    if (root_dir[i] != ' ') {
+      break;
+    }
+    root_dir[i] = '\0';
+  }
+    if (strlen(root_dir) == 0) {
+    ERROR("ROOT_DIR env variable is empty\n");
+    return -1;
+  }
+  snprintf(zl_context.root_dir, sizeof(zl_context.root_dir), "%s", root_dir);
   zl_context.config = *cfg;
 
-  if (chdir(zl_context.workdir)) {
+  zl_context.start_mode = argc > 1 ? ZL_START_MODE_PS : ZL_START_MODE_STC;
```

Comment:
This is a rather hacky way to determine the mode. Is there a better way?

Reply:
The way exactly does what I wanted to. There are two modes: when the launcher starts from `run-zowe.sh` and when the launcher starts independently. I want two distinguish the way by number of command line arguments. Perhaps `ZL_START_MODE_PS` and `ZL_START_MODE_STC` are not very apt names and should be replaced. @ifakhrutdinov  If you have any suggestions please add a comment.

Reply:
I've done some quick research. There seems to be a field in ASCB you could use to distinguish between STC/job/process.

If you follow ASCB->ASCBJBNS, you'll see the job name used with the start command (if I interpreted that correctly).

Let's consider I use the following applications:
* An STC with name `MYSTC`
* A job with name `MYJOB`
* A process `sh test.sh`

||STC|Job|Process|
|-|-|-|-|
|**ASCBJBNS**|MYSTC|INIT|BPXAS|

Even using this method you can check if the value is `BPXAS`, you're a USS process. There may be better way, I'll check with experts and get back to you.

Reply:
I've been told that better ways are:
* Checking `OUCBOMVS` in `OUCBMFL3` (see [OUCB](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.iead300/OUCB-map.htm))
* `IEECHAIN`